libbeat/outputs/shipper/shipper.go

```diff
@@ -70,90 +85,116 @@ func makeShipper(
-func (c *shipper) Connect() error {
-	tls, err := tlscommon.LoadTLSConfig(c.config.TLS)
+func (s *shipper) Connect() error {
+	tls, err := tlscommon.LoadTLSConfig(s.config.TLS)
 	if err != nil {
 		return fmt.Errorf("invalid shipper TLS configuration: %w", err)
 	}
 
 	var creds credentials.TransportCredentials
-	if c.config.TLS != nil && c.config.TLS.Enabled != nil && *c.config.TLS.Enabled {
+	if s.config.TLS != nil && s.config.TLS.Enabled != nil && *s.config.TLS.Enabled {
 		creds = credentials.NewTLS(tls.ToConfig())
 	} else {
 		creds = insecure.NewCredentials()
 	}
 
 	opts := []grpc.DialOption{
 		grpc.WithConnectParams(grpc.ConnectParams{
-			MinConnectTimeout: c.config.Timeout,
+			MinConnectTimeout: s.config.Timeout,
 		}),
 		grpc.WithBlock(),
 		grpc.WithTransportCredentials(creds),
 	}
 
-	ctx, cancel := context.WithTimeout(context.Background(), c.config.Timeout)
+	ctx, cancel := context.WithTimeout(context.Background(), s.config.Timeout)
 	defer cancel()
 
-	c.log.Debugf("trying to connect to %s...", c.config.Server)
+	s.log.Debugf("trying to connect to %s...", s.config.Server)
 
-	conn, err := grpc.DialContext(ctx, c.config.Server, opts...)
+	conn, err := grpc.DialContext(ctx, s.config.Server, opts...)
 	if err != nil {
 		return fmt.Errorf("shipper connection failed with: %w", err)
 	}
 
-	c.conn = conn
-	c.client = sc.NewProducerClient(conn)
+	s.conn = conn
+	s.clientMutex.Lock()
+	defer s.clientMutex.Unlock()
 
-	indexClient, err := c.client.PersistedIndex(ctx, &messages.PersistedIndexRequest{
-		PollingInterval: durationpb.New(0), // no need to subscribe, we need it only once
+	s.client = sc.NewProducerClient(conn)
+
+	// we don't need a timeout context here anymore, we use the
+	// `s.backgroundCtx` instead, it's going to be a long running client
+	ackCtx, ackCanel := context.WithCancel(s.backgroundCtx)
+	defer func() {
+		// in case we return an error before we start the `ackLoop`
+		// then we don't need this client anymore and must close the stream
+		if err != nil {
+			ackCanel()
+		}
+	}()
+
+	indexClient, err := s.client.PersistedIndex(ackCtx, &messages.PersistedIndexRequest{
+		PollingInterval: durationpb.New(s.config.AckPollingInterval),
 	})
 	if err != nil {
 		return fmt.Errorf("failed to connect to the server: %w", err)
 	}
-	serverInfo, err := indexClient.Recv()
+	indexReply, err := indexClient.Recv()
 	if err != nil {
 		return fmt.Errorf("failed to fetch server information: %w", err)
 	}
-	c.serverID = serverInfo.GetUuid()
+	s.serverID = indexReply.GetUuid()
+
+	s.log.Debugf("connection to %s (%s) established.", s.config.Server, s.serverID)
 
-	c.log.Debugf("connection to %s (%s) established.", c.config.Server, c.serverID)
+	go func() {
+		defer ackCanel()
+		s.log.Debugf("starting acknowledgment loop with server %s", s.serverID)
+		// the loop returns only in case of error
+		err := s.ackLoop(s.backgroundCtx, indexClient)
```

Comment:
could this result in ackCancel() being called twice?  Previous defer of ackCancel if err !=nil

Reply:
But this `err` is defined in a new scope (go routine function), it's not the same variable checked in the previous `defer` or am I missing something?

Reply:
I missed the ":", so yeah should be fine.